Source/Core/VideoBackends/D3D12/D3DState.cpp

```diff
@@ -0,0 +1,483 @@
+	}
+	else if (state.min_filter & 4) // linear min filter
+	{
+		if (state.mag_filter) // linear mag filter
+		{
+			if (mip == TexMode0::TEXF_NONE)
+				sampdc.Filter = D3D12_FILTER_MIN_MAG_LINEAR_MIP_POINT;
+			else if (mip == TexMode0::TEXF_POINT)
+				sampdc.Filter = D3D12_FILTER_MIN_MAG_LINEAR_MIP_POINT;
+			else if (mip == TexMode0::TEXF_LINEAR)
+				sampdc.Filter = D3D12_FILTER_MIN_MAG_MIP_LINEAR;
+		}
+		else // point mag filter
+		{
+			if (mip == TexMode0::TEXF_NONE)
+				sampdc.Filter = D3D12_FILTER_MIN_LINEAR_MAG_MIP_POINT;
+			else if (mip == TexMode0::TEXF_POINT)
+				sampdc.Filter = D3D12_FILTER_MIN_LINEAR_MAG_MIP_POINT;
+			else if (mip == TexMode0::TEXF_LINEAR)
+				sampdc.Filter = D3D12_FILTER_MIN_LINEAR_MAG_POINT_MIP_LINEAR;
+		}
+	}
+	else // point min filter
+	{
+		if (state.mag_filter) // linear mag filter
+		{
+			if (mip == TexMode0::TEXF_NONE)
+				sampdc.Filter = D3D12_FILTER_MIN_POINT_MAG_LINEAR_MIP_POINT;
+			else if (mip == TexMode0::TEXF_POINT)
+				sampdc.Filter = D3D12_FILTER_MIN_POINT_MAG_LINEAR_MIP_POINT;
+			else if (mip == TexMode0::TEXF_LINEAR)
+				sampdc.Filter = D3D12_FILTER_MIN_POINT_MAG_MIP_LINEAR;
+		}
+		else // point mag filter
+		{
+			if (mip == TexMode0::TEXF_NONE)
+				sampdc.Filter = D3D12_FILTER_MIN_MAG_MIP_POINT;
+			else if (mip == TexMode0::TEXF_POINT)
+				sampdc.Filter = D3D12_FILTER_MIN_MAG_MIP_POINT;
+			else if (mip == TexMode0::TEXF_LINEAR)
+				sampdc.Filter = D3D12_FILTER_MIN_MAG_POINT_MIP_LINEAR;
+		}
+	}
+
+	sampdc.AddressU = d3d_clamps[state.wrap_s];
+	sampdc.AddressV = d3d_clamps[state.wrap_t];
+	sampdc.AddressW = D3D12_TEXTURE_ADDRESS_MODE_CLAMP;
+
+	sampdc.ComparisonFunc = D3D12_COMPARISON_FUNC_NEVER;
+
+	sampdc.BorderColor[0] = sampdc.BorderColor[1] = sampdc.BorderColor[2] = sampdc.BorderColor[3] = 1.0f;
+
+	sampdc.MaxLOD = (mip == TexMode0::TEXF_NONE) ? 0.0f : (float)state.max_lod / 16.f;
+	sampdc.MinLOD = (float)state.min_lod / 16.f;
+	sampdc.MipLODBias = (s32)state.lod_bias / 32.0f;
+
+	return sampdc;
+}
+
+D3D12_BLEND_DESC StateCache::GetDesc12(BlendState state)
+{
+	if (!state.blend_enable)
+	{
+		state.src_blend = D3D12_BLEND_ONE;
+		state.dst_blend = D3D12_BLEND_ZERO;
+		state.blend_op = D3D12_BLEND_OP_ADD;
+		state.use_dst_alpha = false;
+	}
+
+	D3D12_BLEND_DESC blenddc = {
+		FALSE, // BOOL AlphaToCoverageEnable;
+		FALSE, // BOOL IndependentBlendEnable;
+		{
+			state.blend_enable,   // BOOL BlendEnable;
+			FALSE,                // BOOL LogicOpEnable;
+			state.src_blend,      // D3D12_BLEND SrcBlend;
+			state.dst_blend,      // D3D12_BLEND DestBlend;
+			state.blend_op,       // D3D12_BLEND_OP BlendOp;
+			state.src_blend,      // D3D12_BLEND SrcBlendAlpha;
+			state.dst_blend,      // D3D12_BLEND DestBlendAlpha;
+			state.blend_op,       // D3D12_BLEND_OP BlendOpAlpha;
+			D3D12_LOGIC_OP_NOOP,  // D3D12_LOGIC_OP LogicOp
+			state.write_mask      // UINT8 RenderTargetWriteMask;
+		}
+	};
+
+	if (blenddc.RenderTarget[0].SrcBlend == D3D12_BLEND_SRC_COLOR)
```

Comment:
perhaps using `switch`es here can get rid of some redundancy. or just extract this to its own function like

```cpp
type GetBlendingAlpha(type argument)
{
    switch (argument)
    {
    case D3D12_BLEND_SRC_COLOR:
        return D3D12_BLEND_SRC_ALPHA;
    case D3D12_BLEND_INV_SRC_COLOR:
        return D3D12_BLEND_INV_SRC_ALPHA;
    case D3D12_BLEND_DEST_COLOR:
        return D3D12_BLEND_DEST_ALPHA;
    case D3D12_BLEND_INV_DEST_COLOR:
        return D3D12_BLEND_INV_DEST_ALPHA;
    }

    return argument;
}
```

and then you can just do

```cpp
blenddc.RenderTarget[0].SrcBlendAlpha = GetBlendingAlpha(blenddc.RenderTarget[0].SrcBlend);
blendcc.RenderTarget[0].DestBlendAlpha = GetBlendingAlpha(blenddc.RenderTarget[0].DestBlend);
```